ambry-store/src/main/java/com/github/ambry/store/HybridCompactionPolicy.java

```diff
@@ -0,0 +1,194 @@
+    return selectCompactionPolicy.getCompactionDetails(totalCapacity, usedCapacity, segmentCapacity, segmentHeaderSize,
+        logSegmentsNotInJournal, blobStoreStats, dataDir);
+  }
+
+  /**
+   * Get and Recover the CompactionPolicySwitchInfo from file.
+   * @param dataDir The directory to store the file.
+   * @param storeId id of the BlobStore
+   * @return {@link CompactionPolicySwitchInfo} gets from map or recover from file if needed.
+   */
+  private CompactionPolicySwitchInfo getAndRecoverCompactionPolicySwitchInfo(String storeId, String dataDir) {
+    if (!blobToCompactionPolicySwitchInfoMap.containsKey(storeId)) {
+      File file = new File(Paths.get(dataDir, COMPACT_POLICY_INFO_PATH).toString());
+      if (file.exists()) {
+        CompactionPolicySwitchInfo compactionPolicySwitchInfo = recoverCompactionPolicySwitchInfo(file);
+        blobToCompactionPolicySwitchInfoMap.put(storeId, compactionPolicySwitchInfo);
+      } else {
+        blobToCompactionPolicySwitchInfoMap.put(storeId,
+            new CompactionPolicySwitchInfo(new CompactionPolicyCounter(storeConfig.storeCompactionPolicySwitchPeriod),
+                INIT_COMPACT_ALL_TIME));
+      }
+    }
+    return blobToCompactionPolicySwitchInfoMap.get(storeId);
+  }
+
+  /**
+   * Recover the {@link CompactionPolicySwitchInfo} from backup file.
+   * {
+   *   "compactionPolicyCounter" : {
+   *     "storeCompactionPolicySwitchPeriod" : 3,
+   *     "counter" : 1,
+   *     "value" : 1
+   *   },
+   *   "lastCompactAllTime" : 1593463435900
+   * }
+   * @param file the backup file stores {@link CompactionPolicySwitchInfo}
+   */
+  private CompactionPolicySwitchInfo recoverCompactionPolicySwitchInfo(File file) {
+    try {
+      ObjectMapper objectMapper = new ObjectMapper();
+      return objectMapper.readValue(file, CompactionPolicySwitchInfo.class);
+    } catch (IOException e) {
+      logger.error("Could not deserialize file : {} into {} Object", file, CompactionPolicySwitchInfo.class.getName());
+      return new CompactionPolicySwitchInfo(new CompactionPolicyCounter(storeConfig.storeCompactionPolicySwitchPeriod),
+          INIT_COMPACT_ALL_TIME);
+    }
+  }
+
+  /**
+   * Selects which compaction policy to use for current compaction cycle.
+   * @return {@link CompactionPolicy} to use for current compaction cycle(Selects between {@link StatsBasedCompactionPolicy} and {@link CompactAllPolicy}).
+   */
+  CompactionPolicy selectCompactionPolicyAndUpdateCompactionPolicySwitchInfo(
+      CompactionPolicySwitchInfo compactionPolicySwitchInfo) {
+    if (compactionPolicySwitchInfo == null) {
+      logger.trace("CompactionPolicySwitchInfo is null");
+      return new StatsBasedCompactionPolicy(storeConfig, time);
+    }
+    if (readyToTriggerCompactionAllPolicy(compactionPolicySwitchInfo)) {
+      logger.trace("Return CompactAllPolicy this round");
+      updateCompactionInfo(compactionPolicySwitchInfo);
```

Comment:
Don't you want to do the update either way?

Reply:
I updated the method name to updateCompactionInfoWhenCompactAll to avoid confusion. This method is used to keep record of the last time when we run CompactAllPolicy, so if we passed "storeCompactionPolicySwitchTimestampDays" days since we run compactAllPolicy, we will re-run it again.